### PR TITLE
Remove duplicated code, improve flow

### PR DESCRIPTION
**Pull request by Validark**
- The `pack` function was previously rewritten 4 times, changed to 0
- Made it more obvious that `_queuedResolve` and `_queuedReject` are arrays of functions
- All functionality should be **exactly** the same

Update:
- Removed rightward drift
- Truncated comments to the 80 character column
- Replaced Promise.Status strings to userdata references
- Optimized `andThen` function so it doesn't generate an unnecessary callback function if the Promise already resolved
- Internally differentiated between methods of Promises and the Promise Library itself. This changes `__index` to be its own table. This also means that you can't do this anymore:
	```lua
	Promise.new().new().new().new().new().new()
	```

### CODE REVIEW

File: lib/init.lua [hunk 1]
```diff
@@ -32,8 +30,7 @@ local function wpcallPacked(f, ...)
 	local resultLength, result = pack(xpcall(body, debug.traceback))
 
 	-- If promise debugging is on, warn whenever a pcall fails.
-	-- This is useful for debugging issues within the Promise implementation
-	-- itself.
+	-- This is useful for debugging issues within the Promise implementation itself.
```
Comment (LPGhatguy) [inline lib/init.lua]: Comments are wrapped to 80 columns wide on purpose, which is why this looked kind of strange.

File: lib/init.lua [hunk 2]
```diff
@@ -209,8 +202,8 @@ function Promise:andThen(successHandler, failureHandler)
 
 		if self._status == Promise.Status.Started then
 			-- If we haven't resolved yet, put ourselves into the queue
-			table.insert(self._queuedResolve, successCallback)
-			table.insert(self._queuedReject, failureCallback)
+			self._queuedResolve[#self._queuedResolve + 1] = successCallback
+			self._queuedReject[#self._queuedReject + 1] = failureCallback
```
Comment (LPGhatguy) [inline lib/init.lua]: I don't think this change adds any clarity, and the performance difference is negligible.

File: lib/init.lua [hunk 3]
```diff
@@ -116,7 +109,7 @@ function Promise.new(callback)
 		-- If an error occurs with no observers, this will be set.
 		_unhandledRejection = false,
 
-		-- Queues representing functions we should invoke when we update!
+		-- Queues representing functions we should invoke when we update! (arrays of functions)
```
Comment (LPGhatguy) [inline lib/init.lua]: The way to describe what the type would probably be to change the comment to:

```lua
-- Lists of functions to invoke when our status updates
```

File: lib/init.lua [hunk 4]
```diff
@@ -263,81 +253,71 @@ function Promise:await()
...
-
-		(...):andThen(function(...)
-			self:_resolve(...)
-		end, function(...)
-			self:_reject(...)
-		end)
-
-		return
-	end
+	if self._status == Promise.Status.Started then
```
Comment (LPGhatguy) [inline lib/init.lua]: Early return here reduced rightward drift substantially. Unless we have substantial code that happens in the else-branch, I've favored early-return style for awhile.

File: lib/init.lua [hunk 5]
```diff
@@ -263,81 +253,71 @@ function Promise:await()
...
+		self._status = Promise.Status.Rejected
+		self._valuesLength, self._values = pack(...)
 
-	self._status = Promise.Status.Rejected
-	self._values = {...}
-	self._valuesLength = select("#", ...)
+		local numRejectionHandlers = #self._queuedReject
```
Comment (LPGhatguy) [inline lib/init.lua]: This is a lot clearer. Whatever I was trying to do with `isEmpty` seems silly.

File: lib/init.lua [hunk 6]
```diff
@@ -10,9 +10,7 @@ local PROMISE_DEBUG = false
 	Used to cajole varargs without dropping sparse values.
 ]]
 local function pack(...)
-	local len = select("#", ...)
-
-	return len, { ... }
+	return select("#", ...), { ... }
```
Comment (LPGhatguy) [inline lib/init.lua:13]: Nice, this is a nifty change.

**Review by LPGhatguy — CHANGES_REQUESTED** [review]

**Comment by Validark** [comment]
Updated. Please review again!

**Comment by LPGhatguy** [comment]
Sorry for the long pause on this!

After I merged @evaera's implementation of `Promise.all`, I decided to pick and choose a few things from your PRs. They fix a lot of really really nice things, but also change a few things that I don't think need to be changed.

Take a look at this commit range and let me know if there's anything missing you think is really important: https://github.com/LPGhatguy/roblox-lua-promise/compare/3c9881a478de552445b31916952caa0914360b07...9bda3e68ea927afa2709df7a969af3c06599e5e1

In the future, smaller pull requests would definitely help me merge in formatting fixes like this more easily! I'm going to go ahead and close this one for now.

**Pull request closed without merging**